### PR TITLE
Classify more pucks

### PR DESCRIPTION
**Pull request by SyrianSpock**
I managed to make my code from Saturday night work :tada: however it's using more memory than before :crying_cat_face: . I'm not sure this is the right way to go with the goal/action definitions.

### CODE REVIEW

File: master-firmware/protobuf/strategy.proto [hunk 1]
```diff
@@ -11,18 +11,18 @@ enum PuckColor {
     RED = 0;
     GREEN = 1;
     BLUE = 2;
-    WHITE = 3;
+    REDGREEN = 3;
```
Comment (antoinealb) [inline master-firmware/protobuf/strategy.proto]: nitpick: add comment specifying that it means red or green, aka unknown ?

File: master-firmware/src/strategy/actions.h [hunk 2]
```diff
@@ -6,7 +6,7 @@
 
 /** Number of states goap can visit before giving up. Increasing it means a
  * solution is found on more complex problems at the expense of RAM use. */
-#define GOAP_SPACE_SIZE 40
+#define GOAP_SPACE_SIZE 180
```
Comment (antoinealb) [inline master-firmware/src/strategy/actions.h]: omfg, so expensive.

File: master-firmware/src/strategy/state.cpp [hunk 3]
```diff
@@ -6,10 +6,13 @@ RobotState initial_state(void)
 {
     RobotState state;
     state = RobotState_init_default;
-    for (size_t i = 0; i < 3; i++)
+
+    size_t num_pucks = sizeof(state.puck_available) / sizeof(bool);
```
Comment (antoinealb) [inline master-firmware/src/strategy/state.cpp]: Does this work in a packed context ? Not sure.

File: master-firmware/protobuf/strategy.proto [hunk 4]
```diff
@@ -11,18 +11,18 @@ enum PuckColor {
...
 }
 
 message RobotState {
     option (nanopb_msgopt).msgid = 14;
```
Comment (antoinealb) [inline master-firmware/protobuf/strategy.proto]: I would suggest adding `option (nanopb_msgopt).packed_struct = true;` to RobotState. Using that and reordering the fields to have all the `bool` together (for optimal packing), I got the state size from 32 to 26 bytes. Might not seem much, but every edge matters.

File: master-firmware/src/strategy/goals.h [hunk 5]
```diff
@@ -11,24 +11,31 @@ struct InitGoal : goap::Goal<RobotState> {
     }
 };
 
-struct RedPucksGoal : goap::Goal<RobotState> {
+struct ClassifyStartPucksGoal : goap::Goal<RobotState> {
     virtual int distance_to(const RobotState& state) const
     {
-        return goap::Distance().shouldBeEqual(state.pucks_in_deposit_zone[PuckColor_RED], 2);
+        return goap::Distance().shouldBeEqual(state.classified_pucks[PuckColor_REDGREEN], 3);
```
Comment (antoinealb) [inline master-firmware/src/strategy/goals.h]: Caution, those types of goals will fail completely if the robot can only find 2 pucks for example.

File: master-firmware/protobuf/strategy.proto (hunk 1 above)
Comment (SyrianSpock) [reply master-firmware/protobuf/strategy.proto]: what if I rename it `RED_OR_GREEN`?

File: master-firmware/src/strategy/goals.h (hunk 5 above)
Comment (SyrianSpock) [reply master-firmware/src/strategy/goals.h]: I agree, I'm going to change the goals again over this week

File: master-firmware/src/strategy/state.cpp (hunk 3 above)
Comment (SyrianSpock) [reply master-firmware/src/strategy/state.cpp]: Looking at the generated code, it's a regular C array of booleans
```
bool puck_available[12];
```
so it's fine. Also the `packed=true` doesn't change the generated code, so I guess it's redundant.

**Pull request merged**